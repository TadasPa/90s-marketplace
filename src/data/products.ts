import pictureA from '../assets/images/a.jpg';
import pictureB from '../assets/images/b.jpg';
import IProduct from '../types/product';

export const PRODUCT_A: IProduct = { id: 'a', title: 'Product A', price: '10USD', img: pictureA };
export const PRODUCT_B: IProduct = { id: 'b', title: 'Product B', price: '30USD', img: pictureB };
