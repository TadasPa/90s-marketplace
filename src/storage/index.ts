export const get = (key: string) => window.localStorage.getItem(key) ?? '[]';

export const getObject = <T>(key: string): T => JSON.parse(get(key));

export const set = <T> (key: string, value: T) => window.localStorage.setItem(key, JSON.stringify(value));
