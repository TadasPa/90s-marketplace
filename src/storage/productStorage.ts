import IProduct from "../types/product";
import ICartProduct from "../types/cartProduct";
import { PRODUCTS_KEY } from "./constants";
import { get, getObject, set } from ".";

export const addProduct = (item: IProduct) => { 
    const exists = !!get(PRODUCTS_KEY);
    
    if (exists) {
        const array: ICartProduct[] = getObject<ICartProduct[]>(PRODUCTS_KEY);
        const existingItemIndex = array.findIndex(({ id }) => id === item.id);
        const existingItem = array[existingItemIndex];
        if (existingItemIndex > -1) {
            array[existingItemIndex] = {...existingItem, amount: existingItem.amount + 1 };
        } else {
            array.push({...item, amount: 1});
        }
        set(PRODUCTS_KEY, array);
    } else { 
        set(PRODUCTS_KEY, [{...item, amount: 1}]);
    }
};

export const removeProduct = (item: IProduct) => { 
    const products = getObject<ICartProduct[]>(PRODUCTS_KEY);
    const itemIndex = products.findIndex(({ id }) => id === item.id);
    products.splice(itemIndex, 1);
    set(PRODUCTS_KEY, products);
}

export const getProducts = (): ICartProduct[] => getObject<ICartProduct[]>(PRODUCTS_KEY);