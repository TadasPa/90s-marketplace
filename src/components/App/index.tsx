import { useCallback, useState } from 'react';
import Product from '../../components/Product';
import Cart from '../../components/Cart';
import { addProduct, getProducts, removeProduct } from '../../storage/productStorage';
import { IProduct } from '../../types/product';
import ICartProduct from '../../types/cartProduct';
import { PRODUCT_A, PRODUCT_B } from '../../data/products';

export const App = () => {
  const [cartItems, setCartItems] = useState<ICartProduct[]>(getProducts());
  const totalAmount = cartItems.reduce((total, item) => total += item.amount, 0);
  const handleAdd = useCallback((product: IProduct) => {
    addProduct(product);
    setCartItems(getProducts());
  }, [setCartItems]);

  const handleRemove = useCallback((product: IProduct) => {
    removeProduct(product);
    setCartItems(getProducts());
  }, [setCartItems]);

  return (
    <main>
      <header>
        90s shop
        <nav>
          <ul style={{listStyleType: 'none', display: 'flex'}}>
            <li><a href="/">Home</a></li>
            |
            <li><a href="/cart">Cart ({totalAmount})</a></li>
          </ul>
        </nav>
        <hr/>
      </header>

      {
        window.location.pathname === '/' && (
          <div>
            Welcome to our shop!

            <p>
              You are probably interested in <a href={`/products/${PRODUCT_A.id}`}>{ PRODUCT_A.title}</a>.
            </p>

            <p>
              Check out the newest product <a href={`/products/${PRODUCT_B.id}`}>{ PRODUCT_B.title}</a>!
            </p>
          </div>
        )
      }
      {
        window.location.pathname === `/products/${PRODUCT_A.id}` && (
          <Product product={PRODUCT_A} onAdd={handleAdd} />
        )
      }
      {
        window.location.pathname === `/products/${PRODUCT_B.id}` && (
          <Product product={PRODUCT_B} onAdd={handleAdd} />
        )
      }
      {
        window.location.pathname === '/cart' && (
          <Cart products={cartItems} onRemove={handleRemove} />
        )
      }
    </main>
  );
}

export default App;
