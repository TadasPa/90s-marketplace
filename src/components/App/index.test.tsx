import { render, screen } from "@testing-library/react";
import { LocalStorageMock } from "@react-mock/localstorage";
import App from ".";
import { PRODUCT_A } from "../../data/products";
import { PRODUCTS_KEY } from "../../storage/constants";

const CART_LINK_TESTID = "cart-link";

test("renders App successfully", () => {
  const { container } = render(<App />);
  expect(container).toMatchSnapshot();
});

test("shows cart count from localStorage", async () => {
  render(
    <LocalStorageMock
      items={{ [PRODUCTS_KEY]: JSON.stringify([{ ...PRODUCT_A, amount: 5 }]) }}
    >
      <App />
    </LocalStorageMock>
  );
  expect(await screen.findByTestId(CART_LINK_TESTID)).toHaveTextContent("5");
});
