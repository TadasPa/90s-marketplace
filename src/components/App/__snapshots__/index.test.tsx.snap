// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App successfully 1`] = `
<div>
  <main>
    <header>
      90s shop
      <nav>
        <ul
          style="list-style-type: none; display: flex;"
        >
          <li>
            <a
              data-testid="home-link"
              href="/"
            >
              Home
            </a>
          </li>
          |
          <li>
            <a
              data-testid="cart-link"
              href="/cart"
            >
              Cart (
              0
              )
            </a>
          </li>
        </ul>
      </nav>
      <hr />
    </header>
    <div>
      Welcome to our shop!
      <p>
        You are probably interested in
         
        <a
          href="/products/a"
        >
          Product A
        </a>
        .
      </p>
      <p>
        Check out the newest product
         
        <a
          href="/products/b"
        >
          Product B
        </a>
        !
      </p>
    </div>
  </main>
</div>
`;
